chain-storage/src/store.rs

```diff
@@ -222,42 +222,39 @@ pub trait BlockStore: std::marker::Sized {
     /// Return an iterator that yields block info for the blocks in
     /// the half-open range `(from, to]`. `from` must be an ancestor
     /// of `to` and may be the zero hash.
-    fn iterate_range(
-        &self,
+    fn iterate_range<'store>(
+        &'store self,
         from: &<Self::Block as Block>::Id,
         to: &<Self::Block as Block>::Id,
-    ) -> Result<BlockIterator<Self>, Error> {
+    ) -> Result<BlockIterator<'store, Self::Block>, Error> {
         // FIXME: put blocks loaded by is_ancestor into pending_infos.
         match self.is_ancestor(from, to)? {
             None => panic!(), // FIXME: return error
             Some(distance) => {
                 let to_info = self.get_block_info(&to)?;
                 Ok(BlockIterator {
-                    store: &self,
+                    store: self.as_trait(),
                     to_depth: to_info.depth,
                     cur_depth: to_info.depth - distance,
                     pending_infos: vec![to_info],
                 })
             }
         }
     }
+
+    // See https://stackoverflow.com/questions/42121299/provided-method-casting-self-to-trait-object
+    fn as_trait(&self) -> &BlockStore<Block = Self::Block>;
```

Comment:
Note that there were a few other solutions that don't require a bogus method. I'd go with a helper blanket-implemented trait.

Reply:
you meant something like:

```rust
pub trait BlockStore {
  fn as_trait(&self) -> &BlockStore<Block = Self::Block> { self }
}
```

would that work ?

Reply:
That doesn't work since `self` is not `Sized`, and adding a `Sized` bound to `as_trait` makes it unavailable in trait objects.